### ISSUE TITLE
RFC: Finished states and e-mail notifications.

### ISSUE THREAD

**Issue by siteshwar** [release-engineering/kobo#259]
We are currently trying to add support for [single use workers](https://github.com/openscanhub/openscanhub/pull/234/) in OpenScanHub. However, it is causing issues with the e-mail notifications. E-mail notifications are [sent](https://github.com/release-engineering/kobo/blob/c08f3d9da60998f152a750ac21b6e155bd426651/kobo/worker/taskmanager.py#L464) after the task has moved to the `CLOSED` state. But, `osh-worker-manager` deletes workers as soon as the tasks has moved to the `CLOSED` (or `FAILED`) state, so the e-mail notifications are never sent.

- Shall we send the e-mail notifications before the task reaches the `CLOSED` state? What issues it may cause with the existing deployments? It would require moving [this code block](https://github.com/release-engineering/kobo/blob/c08f3d9da60998f152a750ac21b6e155bd426651/kobo/worker/taskmanager.py#L472-L475) above [changing state](https://github.com/release-engineering/kobo/blob/c08f3d9da60998f152a750ac21b6e155bd426651/kobo/worker/taskmanager.py#L459) of the task.
- If we can not send e-mail notifications before reaching the `CLOSED` state, can we have a separate `notification_sent` field in the `Task` class?
- Is there any other alternative workaround to this situation? I would like to avoid adding `ShutdownWorker` tasks as it would spam the view and the database with `ShutdownWorker` tasks for each task.

**Comment by rohanpm** [comment]
> Shall we send the e-mail notifications before the task reaches the CLOSED state? What issues it may cause with the existing deployments? 

`TaskClass.notification` is not only used for emails.  It's designed to be extended by the user of kobo and it could do anything.  I know for Pub it has long been used to send completion messages on a message bus.

Other services subscribe to those messages, and it would be a significant issue if those services could receive a message, then call back to Pub and find that the relevant task hasn't reached `CLOSED` state.  So, I don't think it can be moved prior to the transition to `CLOSED`.

> If we can not send e-mail notifications before reaching the CLOSED state, can we have a separate notification_sent field in the Task class?

Could you implement this in OSH? If you haven't already found a way, I would really like you to find a way which allows you to extend kobo's models with your own fields, e.g. using inheritance.

> Is there any other alternative workaround to this situation?

Well, from the description it sounds a bit like `osh-worker-manager` is doing the wrong thing.

You mentioned that: `osh-worker-manager` deletes workers as soon as the tasks has moved to the CLOSED (or FAILED) state

From the kobo code (around `finish_task`) it seems pretty clear that the intended control flow for tasks is:

1. task's main body runs
2. then (regardless of success/failure), `TaskClass.cleanup` runs
3. then (regardless of success/failure), `TaskClass.notification` runs

It sounds like you're saying that OSH has somehow arranged things so that step (1) happens but steps (2) and (3) don't happen?  I don't think that would be a problem for kobo to solve, because it seems pretty clear that OSH is "breaking the contract" around how task classes are supposed to work.

Or, worded another way: if you want to do notifications in `TaskClass.notification`, and you customize how tasks are run, then you must ensure to call `TaskClass.notification` when completing a task, like the default taskmanager does.

If `TaskClass.notification` doesn't work for you then you could always just incorporate notifications directly into your task methods, which would cause them to happen before `CLOSED` (and I'm sure you could write them in a way which pretends like the tasks have actually closed, if you want to).  Right now there seems to be an implied position of "We want to implement our email notifications in TaskClass.notification, but in our custom task execution setup we don't want to call that method when completing tasks", it seems like a self-inflicted problem.

**Comment by rohanpm** [comment]
All that being said, I believe the `TaskClass.notification` concept is inherently flawed when you keep in mind that any process or remote service can potentially be interrupted or die without warning:

- if it is called _before_ moving a task to `CLOSED`, then it's possible to send a notification and then be interrupted before the task is actually closed.  Hence you send a notification claiming a task has completed while the API/UI say that it never actually did.
- if it is called _after_ moving a task to `CLOSED`, then it's possible to move the task to `CLOSED` and then be interrupted before notifying.  Hence you permanently miss a notification of a closed task.

In short I'm pretty sure the more correct way for this to be implemented would be: in *the same DB transaction* which sets the task status to `CLOSED`, it also enqueues some kind of "process notifications for this task" message in the DB.  Thus (set task status to CLOSED, commit to sending notifications soon) would be a single atomic operation in the DB.

There is no existing component in kobo capable of processing those kinds of background messages right now, so this would be a rather significant change.  I don't expect anyone has the appetite to work on this, I'm mainly pointing it out for completeness.

**Comment by siteshwar** [comment]
@rohanpm My current idea is to set [enabled](https://github.com/release-engineering/kobo/blob/d4b1c22d9bd35751a909ac40bdd736b3a1e1d54c/kobo/hub/models.py#L191) field in the `Worker` model to `False` in OSH code, once the email notification has been sent. Could it cause any issues if the worker is going to be destroyed after that?

**Comment by rohanpm** [comment]
@siteshwar I'm not quite sure what you mean about possible issues.  As far as I know, OSH is pioneering the idea of deleting workers *at all* (remember it wasn't even possible without also deleting historical tasks prior to  https://github.com/release-engineering/kobo/commit/ea80e349ad120736146228d7b88e37352678d201 ), so you guys would know better than anyone else whether and when deleting workers causes an issue.

**Comment by siteshwar** [comment]
I am closing this issue as it has to be worked around from OpenScanHub side.
